feat: first pass at adding angular signals documentation (#5841)
cypress-io/cypress-documentation: docs/guides/component-testing/angular/examples.mdx

2. [Writable Signal](#Writable-Signal)

##### Inferred Generic Type

In the example below, the `title` prop being passed into our `TestComponent` is a `string`. A `string` is the generic type of our `input()` signal we defined in our `TestComponent`.

```typescript
let titleProp = 'Test Component'
cy.mount(TestComponent, {
  componentProperties: {
    title: titleProp,
  },
})

cy.get('[data-cy="test-component-title-display"]').should(
  'have.text',
  'Test Component'
)
```

:::info
Under the hood, Cypress wraps the generic value in a writable `signal()` and merges it into the prop. In other words, the `@cypress/angular-signals` test harness in this example is really:

```typescript
cy.mount(TestComponent, {
  componentProperties: {
    title: signal('Test Component'),
  },
})
```

:::

This works for any signal. Shown below is an example of testing a `model()` signal with a generic type `number` as seen in our `TestComponent`:

```ts
cy.mount(TestComponent, {
  componentProperties: {
    title: 'Test Component',
    count: 3,
  },
})

cy.get('[data-cy="test-component-count-display"]').should('have.text', '3')
```

##### Writable Signal

> Inferred generic types work very well for most test cases. However, they don't allow us to update the prop in the component after the prop is passed in. For this use case, we need to use a writable `signal()`.

Reviewer: 👍🏻

:::

### Testing Signals

With the releases of Angular versions [17.1](https://github.com/angular/angular/blob/main/CHANGELOG.md#1710-2024-01-17) and [17.2](https://github.com/angular/angular/blob/main/CHANGELOG.md#1720-2024-02-14), [input](https://github.com/angular/angular/pull/53521) and [model](https://github.com/angular/angular/pull/54252) signals were introduced into the `@angular/core` API.

> Since signals introduced new methods and types to the core API, Cypress needed to introduce a new test harness, `@cypress/angular-signals`, in order to support the API without breaking changes.

Reviewer: ```suggestion
Since signals introduced new methods and types to the core API, Cypress introduced a new test harness, `@cypress/angular-signals`.
```

Author: addressed in 1055c7068

Though basic signals were introduced in Angular `16`, this testing harness requires Angular `17.2` and above.

> For our below examples, we will be working with a very simple component called `TestComponent`, which looks something like this:

Reviewer: ```suggestion
For the below examples, we'll be working with a component called `TestComponent`, which looks something like shown below:
```

Author: addressed in 1055c7068

```typescript
// app/components/test-component.component.ts
import { Component, input, model } from '@angular/core'

@Component({
  selector: 'test-component',
  templateUrl: './test-component.component.html',
  standalone: true,
})
export class TestComponent {
  title = input.required<string>()
  count = model.required<number>()
}
```

```html
<!-- app/components/test-component.component.html -->
<p data-cy="test-component-title-display">{{ title() }}</p>
<p data-cy="test-component-count-display">{{ count() }}</p>
<button data-cy="test-component-count-incr" (click)="count.set(count() + 1)">
  Increase
</button>
<button data-cy="test-component-count-decr" (click)="count.set(count() - 1)">
  Decrease
</button>
```

#### Testing Input Signals

> Having `input()` signals work outside the Angular injection context is a bit tricky. Input signals cannot be created outside the Angular injection context, nor is there a way to provide an injection context. Because of this, there are two ways to test `input()` signals with Cypress Component Testing.

Reviewer: What are the 2 ways to test it? I feel like this paragraph drops off without that explanation of where to followup since it merges into all the other sections below without any delineation.

```suggestion
Having `input()` signals work outside of the Angular injection context is a bit tricky. Input signals can't be created outside the Angular injection context, nor is there a way to provide an injection context. Because of this, there are two ways to test `input()` signals within Cypress Component Testing.
```

Author: addressed in 1055c7068

##### Infered Generic Type

> In the example below, the `title` prop being passed into our `TestComponent` is a generic type of the `input()` signal. In this example, that generic type is a `string`. Cypress wraps the generic value in a writable `signal()` and merges it into the prop. In other words:

Reviewer: This is confusion, what's the string type? What's the input signal type? Avoid using 'that' and be specific in the second sentence.

```typescript
let titleProp = 'Test Component'
cy.mount(TestComponent, {
  componentProperties: {
    title: titleProp,
  },
})
```

Under the hood in the `@cypress/angular-signals` test harness is really:

```typescript
cy.mount(TestComponent, {
  componentProperties: {
    title: signal('Test Component'),
  },
})
```

##### Writable Signal

> Since the infered generic type in our example above lives outside the Angular context, there is no way for Cypress to enforce the one-way databinding on context updates. In other words, updates to the `titleProp` variable will not update the prop being passed into `TestComponent`.

Reviewer: ```suggestion
Since the inferred generic type in our example above lives outside of the Angular context, there's no way for Cypress to enforce the one-way data binding on context updates. In other words, updates to the `titleProp` variable will not update the prop being passed into `TestComponent`.
```

Author: addressed in 1055c7068

> To solve this problem, Cypress allows users to pass a writable `signal()` wrapped in the generic type as the prop. This allows us to test our one-way databinding.

Reviewer: ```suggestion
To solve this problem, Cypress allows users to pass a writable `signal()` wrapped in the generic type as the prop. This allows us to test our one-way data binding.
```

Author: addressed in 1055c7068

```typescript
const myTitlePropAsSignal = signal('Test Component')
cy.mount(TestComponent, {
  componentProperties: {
    title: myTitlePropAsSignal,
  },
})

cy.get('the-title-element').should('have.text', 'Test Component')
cy.then(() => {
  // now set the input() through a signal to update the one-way binding
  myTitlePropAsSignal.set('FooBar')
})

cy.get('the-title-element').should('have.text', 'FooBar')
```

###### Change spy for signal `input()`

> Since the prop in question is an `input()`, Cypress does not propagate changes to the `titleChange` spy.

Reviewer: ```suggestion
Since the prop in question is an `input()`, Cypress doesn't propagate changes to the `titleChange` spy.
```

Author: addressed in c084575d0

```typescript
cy.mount(TestComponent, {
  componentProperties: {
    title: 'Test Component',
    // @ts-expect-error
    titleChange: createOutputSpy('titleChange'),
  },
})

// some action that changes the title prop...

// this assertion will not be true. Cypress will never emit change events for props that are `input()` signals

> cy.get('@titleChange').should('have.been.called')

Reviewer: I'd caution against providing examples that don't work. Users inevitably come in, scan, and copy paste then open issues about how the example doesn't work.

```

#### Testing Model Signals

`model()` signals are writable signals, meaning they have the ability to support two-way data binding.

> ##### Infered Generic Type

Reviewer: ```suggestion
##### Inferred Generic Type
```

Author: addressed in 1055c7068

In the example below, the `count` prop being passed into our `TestComponent` is the generic type of the `model()` signal. In this example, that generic type is a `number`. Cypress sets the generic value in the writable `model()`.

```ts
cy.mount(TestComponent, {
  componentProperties: {
    count: 1,
  },
})

cy.get('the-count-element').should('have.text', '1')
```

##### Writable Signal

> Since the infered generic type in our example above lives outside the Angular context, there is no way for Cypress to enforce the databinding on context updates. In other words, updates to the `count` variable will not update the prop being passed into `TestComponent`. This is incredibly similar to how infered generic types of `input()` signals work.

Reviewer: ```suggestion
Since the inferred generic type in our example above lives outside of the Angular context, there's no way for Cypress to enforce the data binding on context updates. In other words, updates to the `count` variable will not update the prop being passed into `TestComponent`. This is incredibly similar to how inferred generic types of `input()` signals work.
```

Author: addressed in 1055c7068

> To solve this problem, Cypress allows users to pass a writable `signal()` wrapped in the generic type as the prop. Since this is a `model()` signal, this allows us to test two-way databinding.

Reviewer: ```suggestion
To solve this problem, Cypress allows users to pass a writable `signal()` wrapped in the generic type as the prop. Since this is a `model()` signal, this allows us to test two-way data binding.
```

Author: addressed in 1055c7068

```typescript
let count = signal(5)
cy.mount(TestComponent, {
  componentProperties: {
    count,
  },
})

// some action that changes the count to 7 inside the component

cy.then(() => {
  // this assertion will be true
  expect(count()).to.be(7)
})
```

```typescript
let count = signal(5)
cy.mount(TestComponentComponent, {
  componentProperties: {
    count,
  },
})

cy.then(() => {
  count.set(8)
})

// this assertion will be true
cy.get('the-count-element').should('have.text', '8')

// later in the test, some action that changes the count to 14 inside the component
// this assertion will be true
cy.get('the-count-element').should('have.text', '14')
```

##### Change spy for models

> Since the prop in question is an `model()` and is a `WritableSignal`, Cypress **will** propagate changes to `countChange` if the output spy is created, either like the example below or if `autoSpyOutputs: true` is configured. However, the typing for `countChange` are not supported since this is technically not a prop (`@Output()` actually adds this as a prop which is not true in our case, even though the signal output is completely valid).

Reviewer: ```suggestion
Since the prop in question is a `model()` and is a `WritableSignal`, Cypress **will** propagate changes to `countChange` if the output spy is created, either like the example below or if `autoSpyOutputs: true` is configured. However, the typing for `countChange` are not supported since this is technically not a prop (`@Output()` actually adds this as a prop which is not true in our case, even though the signal output is completely valid).
```

Author: addressed in 1055c7068
